traffic_ops/traffic_ops_golang/topology/topologies.go

```diff
@@ -0,0 +1,485 @@
+	if err != nil {
+		return nil, fmt.Errorf("topology update: error setting the description for topology %v: %v", topology.Name, err.Error()), http.StatusInternalServerError
+	}
+	for rows.Next() {
+		err = rows.Scan(&topology.Name, &topology.Description, &topology.LastUpdated)
+		if err != nil {
+			return api.ParseDBError(err)
+		}
+	}
+	return nil, nil, http.StatusOK
+}
+
+// Update is a requirement of the api.Updater interface.
+func (topology *TOTopology) Update() (error, error, int) {
+	topologies, userErr, sysErr, errCode := topology.Read()
+	if userErr != nil || sysErr != nil {
+		return userErr, sysErr, errCode
+	}
+	if len(topologies) != 1 {
+		return fmt.Errorf("cannot find exactly 1 topology with the query string provided"), nil, http.StatusBadRequest
+	}
+	oldTopology := TOTopology{APIInfoImpl: topology.APIInfoImpl, Topology: topologies[0].(tc.Topology)}
+	if userErr, sysErr, errCode := topology.setDescription(); userErr != nil || sysErr != nil {
+		return userErr, sysErr, errCode
+	}
+
+	if err := oldTopology.removeParents(); err != nil {
+		return nil, err, http.StatusInternalServerError
+	}
+	var oldNodes, newNodes = map[string]int{}, map[string]int{}
+	for index, node := range oldTopology.Nodes {
+		oldNodes[node.Cachegroup] = index
+	}
+	for index, node := range topology.Nodes {
+		newNodes[node.Cachegroup] = index
+	}
+	var toRemove []string
+	for cachegroupName := range oldNodes {
+		if _, exists := newNodes[cachegroupName]; !exists {
+			toRemove = append(toRemove, cachegroupName)
+		} else {
+			topology.Nodes[newNodes[cachegroupName]].Id = oldTopology.Nodes[oldNodes[cachegroupName]].Id
+		}
+	}
+	if len(toRemove) > 0 {
+		if err := oldTopology.removeNodes(&toRemove); err != nil {
+			return nil, err, http.StatusInternalServerError
+		}
+	}
+	if userErr, sysErr, errCode := topology.addNodes(); userErr != nil || sysErr != nil {
+		return userErr, sysErr, errCode
+	}
+	if userErr, sysErr, errCode := topology.addParents(); userErr != nil || sysErr != nil {
+		return userErr, sysErr, errCode
+	}
+
+	return nil, nil, http.StatusOK
+}
+
+// Delete is unused and simply satisfies the Deleter interface
+// (although TOTOpology is used as an OptionsDeleter)
+func (topology *TOTopology) Delete() (error, error, int) {
+	return nil, nil, 0
+}
+
+// OptionsDelete is a requirement of the OptionsDeleter interface.
+func (topology *TOTopology) OptionsDelete() (error, error, int) {
+	topologies, userErr, sysErr, errCode := topology.Read()
+	if userErr != nil || sysErr != nil {
+		return userErr, sysErr, errCode
+	}
+	if len(topologies) != 1 {
+		return fmt.Errorf("cannot find exactly 1 topology with the query string provided"), nil, http.StatusBadRequest
+	}
+	topology.Topology = topologies[0].(tc.Topology)
+
+	var cachegroups []string
+	for _, node := range topology.Nodes {
+		cachegroups = append(cachegroups, node.Cachegroup)
+	}
+	if err := topology.removeNodes(&cachegroups); err != nil {
```

Comment:
Removing nodes and parents would be handled by the `ON DELETE CASCADE` clause, right? If so, we don't really need to do these steps manually here, but I don't really have a problem doing them here either. I'll leave it up to you to decide :)

Reply:
Yep you're right. Removed in 908e9fb258.